### PR TITLE
SCMOD-6216: Job Service not updating dependent jobs correctly

### PR DESCRIPTION
**Pull request by andyreidz**
@dermot-hardy could you review the SQL function?

### CODE REVIEW

**Comment by davidegph** [comment]
Build is passing
http://sou-jenkins2.hpeswlab.net/job/JobService/view/Developer/job/JobService%7Ejob-service%7ESCMOD-6216-Andy%7ECI/

File: job-service-db/src/main/resources/procedures/processDependentJobs.sql [hunk 1]
```diff
@@ -33,68 +33,56 @@ RETURNS TABLE(
...
-                SELECT NULL
-                FROM job_dependency jd2
-                INNER JOIN job j2
-                    ON j2.job_Id = jd2.dependent_job_id
-                WHERE jd2.job_id = jd.job_id
-                    AND jd2.dependent_job_id <> in_job_id
-                    AND j2.status <> 'Completed');
+        SELECT j.job_id, j.delay
+		FROM job_dependency jd
+		LEFT JOIN job j ON j.job_id = jd.job_id
```
Comment (dermot-hardy) [inline job-service-db/src/main/resources/procedures/processDependentJobs.sql]: This should be an `INNER JOIN`, and I'd tend to put it the other way around but that's probably just a matter of preference.

File: job-service-db/src/main/resources/procedures/processDependentJobs.sql [hunk 2]
```diff
@@ -33,68 +33,56 @@ RETURNS TABLE(
...
+		WHERE jd.dependent_job_id = in_job_id;
 
-    RETURN QUERY
-    SELECT * FROM tmp_dependent_jobs_ready_to_run;
+    PERFORM NULL FROM public.job WHERE job.job_id IN (SELECT tmp_dependent_jobs.job_id FROM tmp_dependent_jobs) ORDER BY job.job_id ASC FOR UPDATE;
```
Comment (dermot-hardy) [inline job-service-db/src/main/resources/procedures/processDependentJobs.sql]: - If this lock is needed then I think it should be in the first statement.
- However I can't quite think why it is needed. (I'll think about it more shortly.)

File: job-service-db/src/main/resources/procedures/processDependentJobs.sql [hunk 3]
```diff
@@ -33,68 +33,56 @@ RETURNS TABLE(
...
-    UPDATE job_task_data jtd
-    SET eligible_to_run_date = now() AT TIME ZONE 'UTC' + (j.delay * interval '1 second')
-    FROM job j
-    INNER JOIN tmp_dependent_jobs_with_delay tdjwd
-        ON tdjwd.job_id = j.job_id
-    WHERE jtd.job_id = j.job_id;
+    --Return jobs with no delay that we can now run
+    RETURN QUERY SELECT jtd.job_id, jtd.task_classifier, jtd.task_api_version, jtd.task_data, jtd.task_pipe, jtd.target_pipe
+    FROM job_task_data jtd
+        LEFT JOIN tmp_dependent_jobs dp ON dp.job_id = jtd.job_id
```
Comment (dermot-hardy) [inline job-service-db/src/main/resources/procedures/processDependentJobs.sql]: I think this should be `INNER JOIN`.

File: job-service-db/src/main/resources/procedures/processDependentJobs.sql [hunk 4]
```diff
@@ -33,68 +33,56 @@ RETURNS TABLE(
...
+        FROM  job_task_data jtd 
+        WHERE jtd.job_id IN (
+			SELECT jtd.job_id
+			    FROM job_task_data jtd
+        			LEFT JOIN tmp_dependent_jobs dp ON dp.job_id = jtd.job_id	
+    			WHERE dp.delay = 0 AND NOT EXISTS (
+        			SELECT job_dependency.job_id 
+						FROM job_dependency 
+            			WHERE job_dependency.job_id = dp.job_id
+    ));
```
Comment (dermot-hardy) [inline job-service-db/src/main/resources/procedures/processDependentJobs.sql]: These two statements should be combined with a `RETURNING *`.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

File: job-service-db/src/main/resources/procedures/processDependentJobs.sql (hunk 2 above)
Comment (dermot-hardy) [reply job-service-db/src/main/resources/procedures/processDependentJobs.sql]: As discussed, this can deadlock with a new job at the minute.

Comment (andyreidz) [reply job-service-db/src/main/resources/procedures/processDependentJobs.sql]: Commented out this statement to allow merges to proceed.

Comment (dermot-hardy) [reply job-service-db/src/main/resources/procedures/processDependentJobs.sql]: I've raised [SCMOD-6620](https://portal.digitalsafe.net/browse/SCMOD-6620) for this.

**Pull request merged**